mkdocs/docs/bindings/c-api.md

```diff
@@ -1,9 +1,217 @@
+API header files are organized by runtime component:
+
+| Component header file                                                       | Overview                                                                  |
+|-----------------------------------------------------------------------------|---------------------------------------------------------------------------|
+| [iree/base/api.h](https://github.com/google/iree/blob/main/iree/base/api.h) | Core API, type definitions, ownership policies, utilities                 |
+| [iree/vm/api.h](https://github.com/google/iree/blob/main/iree/vm/api.h)     | VM APIs: loading modules, I/O, calling functions                          |
+| [iree/hal/api.h](https://github.com/google/iree/blob/main/iree/hal/api.h)   | HAL APIs: device management, synchronization, accessing hardware features |
+
+The [iree/samples/](https://github.com/google/iree/tree/main/iree/samples)
+directory demonstrates several ways to use IREE's C API.
 
 ## Prerequisites
 
-<!-- TODO(??): build from source, vcpkg -->
+To use IREE's C API, you will need to build the runtime
+[from source](../building-from-source/getting-started.md). The
+[iree-template-cpp](https://github.com/iml130/iree-template-cpp) community
+project also shows how to integrate IREE into an external project using
+CMake[^2].
+
+## Concepts
+
+### IREE VM
+
+<!-- TODO(scotttodd): VM "instances" -->
+<!-- TODO(scotttodd): VM "contexts" -->
+<!-- TODO(scotttodd): VM "modules" -->
+
+### IREE HAL
+
+<!-- TODO(scotttodd): HAL "drivers" -->
+<!-- TODO(scotttodd): HAL "devices" -->
+
+## Using the C API
+
+<!-- TODO(scotttodd): general overview (setup, prepare input, call function, get output) -->
+
+### Setup
+
+Include headers:
+
+``` c
+#include "iree/base/api.h"
+#include "iree/hal/api.h"
+#include "iree/vm/api.h"
+
+// The VM and HAL native modules will typically be included, along
+// with those for the specific HAL drivers your application uses.
+// Functionality extensions can be used via custom modules.
+#include "iree/modules/hal/hal_module.h"
+#include "iree/hal/dylib/registration/driver_module.h"
+#include "iree/vm/bytecode_module.h"
+```
+
+Check the API version and register components:
+
+``` c
+// Statically linking the runtime should never have version conflicts,
+// however dynamically linking against a runtime shared object must
+// always verify that the version is as expected.
+iree_api_version_t actual_version;
+IREE_CHECK_OK(iree_api_version_check(IREE_API_VERSION_LATEST, &actual_version));
+
+// Modules with custom types must be statically registered before use.
+IREE_CHECK_OK(iree_hal_module_register_types());
+
+// Device drivers are managed through registries.
+// Applications may use multiple registries to more finely control driver
+// lifetimes and visibility.
+IREE_CHECK_OK(iree_hal_dylib_driver_module_register(
+    iree_hal_driver_registry_default()));
+```
+
+!!! tip
+    Macros like `IREE_CHECK_OK()` can be used to help with error handling.
```

Comment:
IREE_CHECK_OK() may relying on the underlying iree/base:logging library; IREE_RETURN_IF_ERROR() has only the dependency of iree/base:api and still provides the context of the error (with the default IREE_STATUS_MODE=2).

Reply:
`IREE_CHECK_OK` is defined in the same header and easier to understand. Logging [can be, will be able to be] stripped. `IREE_RETURN_IF_ERROR` requires changing your existing code to return `iree_status_t`.